File: SwiftTestApp/SwiftTestApp/LocationManager.swift
//
//  LocationManager.swift
//  SwiftTestApp
//
//  Created by GiPStech on 25/03/25.
//

import SwiftUI
import GiPStechMDK

class LocationManager: ObservableObject {
    let DEV_KEY = // Your Development Key
    let BUILDING_ID = // Your Building Id
    let UNIVERSAL = true // Set to false for indoor localization

    var session: LocationSession?

    @Published var levelText: String = ""
    @Published var geofenceText: String = ""
    @Published var locationText: String = "Ready"

    public func report(_ message: String) {
        DispatchQueue.main.async {
            self.locationText = message
        }
    }

    public func reportLevel(_ message: String) {
        DispatchQueue.main.async {
            self.levelText = message
        }
    }

    public func reportGeofence(_ message: String) {
        DispatchQueue.main.async {
            self.geofenceText = message
        }
    }

    func restartLocalization() async throws {
        do {
            let locationListener = MyLocationListener(manager: self)
            session = try await session?.target.startLocalization(listener: locationListener)
        } catch let error as NSError {
            report("Error: \(error.localizedDescription)")
        }
    }

    func startLocalization() {
        Task {
            do {
                if !(GiPStech.shared.isInitialized) {
                    _ = try await GiPStech.shared.doInit(devKey: DEV_KEY)
                }

                if UNIVERSAL {
                    try await startUniversalLocalization()
                } else {
                    try await startIndoorLocalization()
                }
            } catch let error as NSError {
                let ex = error.kotlinException as? GiPStechException
                report("Error: \(ex?.message ?? "Unknown error")")
            }
        }
    }

    func startIndoorLocalization() async throws {
        let building = try await SpatialManager.shared.requestBuilding(buildingId: BUILDING_ID, fallback: true)
        let locationListener = MyLocationListener(manager: self)
        try await locationListener.onBuildingChanged(building: building)

        session = try await building.startLocalization(listener: locationListener)
    }

    func startUniversalLocalization() async throws {
        if let location = try await SpatialManager.shared.getLocationFromOS() {
            let pair = location.getBoundingBox(size: 100)
            let region = try await SpatialManager.shared.requestRegion(southWest: pair.first!, northEast: pair.second!, fallback: true)
            let locationListener = MyLocationListener(manager: self)

            session = try await region.startLocalization(listener: locationListener)
        }
    }

    func stopLocalization() {
        Task {
            try await session?.target.stopLocalization()
        }
    }
}

File: KmmTestApp/composeApp/src/commonMain/kotlin/com/gipstech/kmmtestapp/LocationManager.kt
package com.gipstech.kmmtestapp

import androidx.compose.runtime.mutableStateOf
import com.gipstech.mdk.GiPStech
import com.gipstech.mdk.LocationSession
import com.gipstech.mdk.SpatialManager
import com.gipstech.mdk.spatial.Place
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

object LocationManager {
    private const val DEV_KEY = // Your Development Key
    private const val BUILDING_ID = // Your Building Id
    private const val UNIVERSAL = true // Set to false for indoor localization

    var place: Place? = null
    val isButtonEnabled = mutableStateOf(false)
    val levelMessage = mutableStateOf("")
    val locationMessage = mutableStateOf("Starting...")
    val geofenceMessage = mutableStateOf("")
    private var session: LocationSession? = null
    private val activityScope = CoroutineScope(Dispatchers.Main)

    fun init(context: Any?) {
        activityScope.launch {
            try {
                GiPStech.init(DEV_KEY, null, context)
                locationMessage.value = "Ready"
                isButtonEnabled.value = true
            } catch (e: Exception) {
                locationMessage.value = "Error: ${e.message}"
            }
        }
    }

    fun startLocalization() {
        levelMessage.value = ""
        locationMessage.value = ""
        geofenceMessage.value = ""

        activityScope.launch {
            try {
                if (UNIVERSAL) {
                    startUniversalLocalization()
                } else {
                    startIndooLocalization()
                }
            } catch (e: Exception) {
                locationMessage.value = "Error: ${e.message}"
                isButtonEnabled.value = false
            }
        }
    }

    private suspend fun startIndooLocalization() {
        val building = SpatialManager.requestBuilding(BUILDING_ID)
        val listener = MyLocationListener()
        listener.onBuildingChanged(building)

        place = building
        session = building.startLocalization(listener)
    }

    private suspend fun startUniversalLocalization() {
        val location = SpatialManager.getLocationFromOS()

        if (location != null) {
            val (min, max) = location.getBoundingBox(1000.0)
            val region = SpatialManager.requestRegion(min, max)
            val listener = MyLocationListener()

            place = region
            session = region.startLocalization(listener)
        } else {
            locationMessage.value = "Position not available"
        }
    }

    fun stopLocalization() {
        activityScope.launch {
            try {
                session!!.target.stopLocalization()
            } catch (e: Exception) {
                locationMessage.value = "Error: ${e.message}"
                isButtonEnabled.value = false
            }
        }
    }
}
